checks/doks/node_labels_taints.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2019 DigitalOcean
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package doks
+
+import (
+	"strings"
+
+	"github.com/digitalocean/clusterlint/checks"
+	"github.com/digitalocean/clusterlint/kube"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func init() {
+	checks.Register(&nodeLabelsTaintsCheck{})
+}
+
+type nodeLabelsTaintsCheck struct{}
+
+// Name returns the name of the check.
+func (*nodeLabelsTaintsCheck) Name() string {
+	return "node-labels-and-taints"
+}
+
+// Groups returns groups for this check.
+func (*nodeLabelsTaintsCheck) Groups() []string {
+	return []string{"doks"}
+}
+
+// Description returns a description of the check.
+func (*nodeLabelsTaintsCheck) Description() string {
+	return "Checks that nodes do not have custom labels or taints configured."
+}
+
+// Run runs the check.
+func (c *nodeLabelsTaintsCheck) Run(objects *kube.Objects) ([]checks.Diagnostic, error) {
+	var diagnostics []checks.Diagnostic
+
+	for _, node := range objects.Nodes.Items {
+		for labelKey := range node.Labels {
+			if !isKubernetesLabel(labelKey) && !isDOKSLabel(labelKey) {
+				d := checks.Diagnostic{
+					Check:    c.Name(),
+					Severity: checks.Warning,
```

Comment:
Should it be a warning or error? If custom labels are used as node selectors in pod specs, it could be a breaking change, right?

Reply:
I went with a warning here since the custom labels/taints won't break an upgrade (the upgrade itself will work fine, but workloads may be affected), and there are probably some valid reasons to use custom labels/taints in DOKS despite their non-persistence. For example, a user may have some kind of controller running that automatically labels/taints nodes continuously so it's fine for them to go away as part of upgrade.